osquery/tables/system/windows/windows_update_history.cpp

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <string>
+
+#include <atlbase.h>
+#include <windows.h>
+#include <wuapi.h>
+
+#include <osquery/core/core.h>
+#include <osquery/core/tables.h>
+#include <osquery/logger/logger.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/conversions/windows/windows_time.h>
+
+namespace osquery {
+namespace tables {
+
+HRESULT populateRow(IUpdateHistoryEntry* entry, Row& r);
+
+QueryData genUpdateHistory(QueryContext& context) {
+  QueryData results;
+
+  CComPtr<IUpdateSearcher> updateSearcher;
+  HRESULT hr = updateSearcher.CoCreateInstance(CLSID_UpdateSearcher);
+  if (FAILED(hr)) {
+    TLOG << "Failed to instantiate IUpdateSearcher";
+    return results;
+  }
+
+  LONG totalCount = 0;
+  hr = updateSearcher->GetTotalHistoryCount(&totalCount);
+  if (FAILED(hr)) {
+    TLOG << "Failed to get total update history count";
+    return results;
+  }
+
+  CComPtr<IUpdateHistoryEntryCollection> updateHistoryEntryCollection;
+  hr = updateSearcher->QueryHistory(
+      0, totalCount, &updateHistoryEntryCollection);
+  if (FAILED(hr)) {
+    TLOG << "Failed to query update history";
+    return results;
+  }
+
+  LONG count = 0;
+  hr = updateHistoryEntryCollection->get_Count(&count);
+  if (FAILED(hr)) {
+    TLOG << "Failed to get update history entry collection size";
+    return results;
+  }
+
+  for (LONG i = 0; i < count; i++) {
+    CComPtr<IUpdateHistoryEntry> entry;
+    hr = updateHistoryEntryCollection->get_Item(i, &entry);
+    if (FAILED(hr)) {
+      TLOG << "Failed to get update history entry";
+      return results;
+    }
+    Row r;
+    hr = populateRow(entry, r);
+    if (FAILED(hr)) {
+      TLOG << "Failed to populate result row from update history entry";
+      return results;
+    }
+    results.push_back(std::move(r));
+  }
+
+  return results;
+}
+
+HRESULT populateRow(IUpdateHistoryEntry* entry, Row& r) {
+  HRESULT hr = S_OK;
+
+  CComBSTR appID;
+  hr = entry->get_ClientApplicationID(&appID);
+  if (FAILED(hr)) {
+    return hr;
+  }
+  r["client_app_id"] = bstrToString(appID);
+
+  DATE date = 0;
```

Comment:
If possible, variables should be declared near the use site, and should always be initialized. Example: `SYSTEMTIME st{};`